### PR TITLE
Allow hashes to be calculated based on non-installed DESCRIPTION files

### PR DESCRIPTION
**Pull request by jcheng5**
Related:
https://github.com/rstudio/connect/pull/514 (see here for summary)
https://github.com/rstudio/rsconnect/pull/42

### CODE REVIEW

**Comment by jcheng5** [comment]
If we want to call `.hash` from rsconnect, maybe it should be renamed to `.hash` and exported?

**Comment by aronatkins** [comment]
Bump the package version. :+1: otherwise.

I'm OK with renaming `packrat:::hash` to `packrat::hash` but don't see why you prefer `.hash`.

**Comment by aronatkins** [comment]
After mulling this a bit...

Maybe `hash` should be based off the `DESCRIPTION` data-frame rather than a path to a `DESCRIPTION` file; then we get the ability to change how we obtain that description data without further changing packrat (and pushing new versions to CRAN).

**Comment by kevinushey** [comment]
Why does `hash` need to be exported? Can't you just call it with `:::`? (Or is this to avoid R CMD check grumpiness?)

File: R/cache.R [hunk 1]
```diff
@@ -7,17 +7,23 @@ isUsingCache <- function(project) {
...
 }
 
+installedDescLookup <- function(pkgName) {
+  system.file("DESCRIPTION", package = pkgName)
+}
+
 # We assume 'path' is the path to a DESCRIPTION file
 #' @importFrom tools md5sum
-hash <- function(path) {
+hash <- function(path, descLookup = NULL) {
```
Comment (kevinushey) [inline R/cache.R]: Why not just use `descLookup = installedDescLookup`?

**Comment by kevinushey** [comment]
FWIW, if you want to 'sneakily' use a non-exported function and dodge R CMD check, you can probably use something like:

```R
extract_from <- function(package, object) {
    do.call(":::", list(as.name(package), as.name(object)))
}

hash <- extract_from("packrat", "hash")
```

Although maybe just exporting `.hash` has a (hidden) export would be fine.

**Comment by jcheng5** [comment]
I'm fine with hash not being exported. (Not going to CRAN with the calling code) We have a ::.snapshotImpl that's called from the same place in connect, so I was wondering if I should follow that pattern.

**Comment by kevinushey** [comment]
My nits early are minor enough that I'm happy merging as-is. Anything else you wanted to change?

**Comment by jcheng5** [comment]
Yeah, one more round tonight to incorporate Aron's feedback.

**Comment by jcheng5** [comment]
@kevinushey @aronatkins Ready for review again.

@kevinushey Do you mind bumping the version right after the merge, so @aronatkins can update Connect's minimum packrat version requirement?

**Comment by aronatkins** [comment]
:mouse: I'm good.

**Comment by jcheng5** [comment]
@kevinushey Never mind, I'm on it

**Pull request merged**